StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c

```diff
@@ -8,6 +8,127 @@
 
 #include "StandaloneMmIplPei.h"
 
+EFI_PEI_MM_COMMUNICATION_PPI  mMmCommunicationPpi = { Communicate };
+
+EFI_PEI_PPI_DESCRIPTOR  mPpiList = {
+  (EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST),
+  &gEfiPeiMmCommunicationPpiGuid,
+  &mMmCommunicationPpi
+};
+
+/**
+  Communicates with a registered handler.
+
+  This function provides a service to send and receive messages from a registered UEFI service.
+
+  @param[in] This                The EFI_PEI_MM_COMMUNICATION_PPI instance.
+  @param[in, out] CommBuffer     A pointer to the buffer to convey into MMRAM.
+  @param[in, out] CommSize       The size of the data buffer being passed in.On exit, the size of data
+                                 being returned. Zero if the handler does not wish to reply with any data.
+
+  @retval EFI_SUCCESS            The message was successfully posted.
+  @retval EFI_INVALID_PARAMETER  The CommBuffer was NULL.
+  @retval EFI_NOT_STARTED        The service is NOT started.
+**/
+EFI_STATUS
+EFIAPI
+Communicate (
+  IN CONST EFI_PEI_MM_COMMUNICATION_PPI  *This,
+  IN OUT VOID                            *CommBuffer,
+  IN OUT UINTN                           *CommSize
+  )
+{
+  EFI_STATUS                  Status;
+  PEI_SMM_CONTROL_PPI         *SmmControl;
+  UINT8                       SmiCommand;
+  UINTN                       Size;
+  EFI_SMM_COMMUNICATE_HEADER  *CommunicateHeader;
+  UINTN                       TempCommSize;
+  EFI_HOB_GUID_TYPE           *GuidHob;
+  MM_COMM_BUFFER              *MmCommBuffer;
+  COMMUNICATION_IN_OUT        *CommunicationInOut;
+
+  DEBUG ((DEBUG_INFO, "StandaloneMmIpl Communicate Enter\n"));
+
+  GuidHob = GetFirstGuidHob (&gEdkiiCommunicationBufferGuid);
+  if (GuidHob != NULL) {
+    MmCommBuffer       = GET_GUID_HOB_DATA (GuidHob);
+    CommunicationInOut = (COMMUNICATION_IN_OUT *)(UINTN)MmCommBuffer->CommunicationInOut;
+  } else {
+    DEBUG ((DEBUG_ERROR, "MmCommBuffer is not existed !!!\n"));
+    ASSERT (GuidHob != NULL);
+    return EFI_NOT_FOUND;
+  }
+
+  SmiCommand = 0;
+  Size       = sizeof (SmiCommand);
+
+  //
+  // Check parameters
+  //
+  if (CommBuffer == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  CommunicateHeader = (EFI_SMM_COMMUNICATE_HEADER *)CommBuffer;
+
+  if (CommSize == NULL) {
+    TempCommSize = OFFSET_OF (EFI_SMM_COMMUNICATE_HEADER, Data) + CommunicateHeader->MessageLength;
+  } else {
+    TempCommSize = *CommSize;
+    //
+    // CommSize must hold Header length and MessageLength
+    //
+    if (TempCommSize != (OFFSET_OF (EFI_SMM_COMMUNICATE_HEADER, Data) + CommunicateHeader->MessageLength)) {
+      return EFI_INVALID_PARAMETER;
+    }
+  }
+
+  if (TempCommSize > MmCommBuffer->FixedCommBufferSize) {
+    DEBUG ((DEBUG_ERROR, "Communicate buffer size is over MAX size\n"));

```

Comment:
"Communicate buffer size (%d) is over MAX (%d) size!", TempCommSize, MmCommBuffer->FixedCommBufferSize